cpp/include/raft/sparse/matrix/detail/preprocessing.cuh

```diff
@@ -0,0 +1,463 @@
+                  origin_map.data_handle());
+  thrust::for_each(raft::resource::get_thrust_policy(handle),
+                   result.data_handle(),
+                   result.data_handle() + result.size(),
+                   mapper<T2>(raft::make_const_mdspan(origin_map.view())));
+}
+
+/**
+ * @brief Compute row(id) counts
+ * @param handle: raft resource handle
+ * @param rows: Input COO rows array
+ * @param columns: Input COO columns array
+ * @param values: Input COO values array
+ * @param id_counts: Output array that stores counts per row, scattered to same shape as rows.
+ * @param n_rows: Number of rows in matrix
+ */
+template <typename T1, typename T2, typename IdxT>
+void get_id_counts(raft::resources& handle,
+                   raft::device_vector_view<T1, IdxT> rows,
+                   raft::device_vector_view<T1, IdxT> columns,
+                   raft::device_vector_view<T2, IdxT> values,
+                   raft::device_vector_view<T2, IdxT> id_counts,
+                   T1 n_rows)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+
+  auto row_keys   = raft::make_device_vector<T1, IdxT>(handle, n_rows);
+  auto row_counts = raft::make_device_vector<T2, IdxT>(handle, n_rows);
+  auto row_fill   = raft::make_device_vector<T2, IdxT>(handle, rows.size());
+
+  // the amount of columns(features) that each row(id) is found in
+  thrust::fill(raft::resource::get_thrust_policy(handle),
+               row_fill.data_handle(),
+               row_fill.data_handle() + row_fill.size(),
+               1.0f);
+  get_uniques_counts(
+    handle, rows, columns, values, row_fill.view(), row_keys.view(), row_counts.view());
+
+  create_mapped_vector<T1, T2>(handle, rows, row_keys.view(), row_counts.view(), id_counts);
+}
+
+/**
+ * @brief Gather per feature mean values, returns the cumulative avg feature length.
+ * @param handle: raft resource handle
+ * @param rows: Input COO rows array
+ * @param columns: Input COO columns array
+ * @param values: Input COO values array
+ * @param feat_lengths: Output array that stores mean per feature value
+ * @param n_cols: Number of columns in matrix
+ */
+template <typename T1, typename T2, typename IdxT>
+int get_feature_data(raft::resources& handle,
+                     raft::device_vector_view<T1, IdxT> rows,
+                     raft::device_vector_view<T1, IdxT> columns,
+                     raft::device_vector_view<T2, IdxT> values,
+                     raft::device_vector_view<T2, IdxT> feat_lengths,
+                     T1 n_cols)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+
+  auto col_keys   = raft::make_device_vector<T1, IdxT>(handle, n_cols);
+  auto col_counts = raft::make_device_vector<T2, IdxT>(handle, n_cols);
+
+  get_uniques_counts(handle, columns, rows, values, values, col_keys.view(), col_counts.view());
+  int total_feature_lengths = thrust::reduce(raft::resource::get_thrust_policy(handle),
```

Comment:
We have raft primitives for this- please use those. The reason we prefer RAFT primitives where at all possible is because we have direct control over RAFT but not Thrust. When thrust changes, or we find a perf issue in Thrust, we'll need to go through the whole codebase and change ever occurrence. If you rely on RAFT primitives, we centralize all those calls so we only have to change one place.